File: packages/sdk/src/sources/grok/sidecars.ts
/**
 * Grok sidecar enrichment — timestamps from events.jsonl, session tokens from
 * signals.json.
 *
 * ## Timestamp join (turn-scoped — verified against real ~/.grok sessions)
 *
 * `chat_history.jsonl` has no per-message time. `events.jsonl` does, with a
 * reliable structure:
 *
 * 1. **`turn_started.conversation_message_count`** is the absolute
 *    chat_history line index of that turn's primary user message (exact match
 *    on every observed turn).
 * 2. Turn ranges are `[count_i, count_{i+1})` (last turn → EOF).
 * 3. Within a turn's time window `[turn_started.ts, turn_ended.ts]`:
 *    - `loop_started[]` and `first_token[]` are 1:1 with **assistant** cycles
 *      (not with reasoning — a loop may emit multiple reasoning records before
 *      one assistant).
 *    - Walk chat lines in order:
 *      - `user` → `turn_started.ts`
 *      - `reasoning` → current `loop_started[loop_i].ts` (stay on same loop)
 *      - `assistant` → `first_token[loop_i].ts`, then `loop_i++`
 * 4. Pre-turn lines (`0 .. first_count`) — system + bootstrap context users —
 *    get `fallbackCreated` (summary.created_at).
 *
 * Without events.jsonl, only the fallback is applied to system/user lines.
 *
 * ## Session tokens
 *
 * `signals.contextTokensUsed` is a session aggregate. Attribute it to the last
 * assistant message and set `tokens_estimated=1` so the UI never treats it as
 * per-message API usage.
 */

import * as path from 'node:path';

import type { FileService } from '../../io/file-service.js';
import type { SessionTokenApi } from '../types.js';

const EVENTS_FILE = 'events.jsonl';
const SIGNALS_FILE = 'signals.json';

export interface GrokSignals {
  contextTokensUsed: number;
  contextWindowTokens: number;
}

/** Minimal event shape we care about for timestamp attribution. */
export interface GrokEventLine {
  type: string;
  ts: string;
  conversation_message_count?: number;
  turn_number?: number;
  loop_index?: number;
}

/**
 * Build absolute-line-index → ISO timestamp map from chat line types + events.
 * `lineTypes[i]` is the `type` of non-empty chat_history line i (including
 * tool lines that do not become message rows).
 */
export function buildTimestampMap(
  lineTypes: string[],
  events: GrokEventLine[],
  fallbackCreated: string | null,
): Map<number, string> {
  const map = new Map<number, string>();
  const n = lineTypes.length;
  if (n === 0) return map;

  const turns = events
    .filter((e) => e.type === 'turn_started' && e.ts)
    .slice()
    .sort((a, b) => {
      const ta = a.turn_number ?? -1;
      const tb = b.turn_number ?? -1;
      if (ta >= 0 && tb >= 0 && ta !== tb) return ta - tb;
      return a.ts < b.ts ? -1 : a.ts > b.ts ? 1 : 0;
    });

  const turnEnds = events.filter((e) => e.type === 'turn_ended' && e.ts).map((e) => e.ts);

  // ── Pre-turn bootstrap (system + context users before first turn) ────────
  const firstTurnStart = turns.length > 0 ? (turns[0].conversation_message_count ?? 0) : n;
  const preEnd = clampIndex(firstTurnStart, 0, n);
  if (fallbackCreated) {
    for (let i = 0; i < preEnd; i++) {
      const t = lineTypes[i];
      if (t === 'system' || t === 'user') {
        map.set(i, fallbackCreated);
      }
    }
  }

  if (turns.length === 0) {
    // No turn markers — apply fallback to remaining system/user only.
    if (fallbackCreated) {
      for (let i = preEnd; i < n; i++) {
        if (lineTypes[i] === 'system' || lineTypes[i] === 'user') {
          map.set(i, fallbackCreated);
        }
      }
    }
    return map;
  }

  // ── Per-turn scoped join ─────────────────────────────────────────────────
  for (let ti = 0; ti < turns.length; ti++) {
    const turn = turns[ti];
    const start = clampIndex(turn.conversation_message_count ?? 0, 0, n);
    const end = ti + 1 < turns.length ? clampIndex(turns[ti + 1].conversation_message_count ?? n, start, n) : n;

    const windowStart = turn.ts;
    // Prefer matching turn_ended by order; else next turn_started; else open-ended.
    const windowEnd = ti < turnEnds.length ? turnEnds[ti] : ti + 1 < turns.length ? turns[ti + 1].ts : '\uffff';

    const loops = events.filter((e) => e.type === 'loop_started' && e.ts && e.ts >= windowStart && e.ts <= windowEnd);
    const firstTokens = events.filter(
      (e) => e.type === 'first_token' && e.ts && e.ts >= windowStart && e.ts <= windowEnd,
    );

    let loopI = 0;
    for (let i = start; i < end; i++) {
      const t = lineTypes[i];
      if (t === 'user' || t === 'system') {
        // Primary user is at `start`; extra users in the same turn share turn_started.
        map.set(i, turn.ts);
      } else if (t === 'reasoning') {
        // Multiple reasonings may precede one assistant within the same loop.
        if (loopI < loops.length) {
          map.set(i, loops[loopI].ts);
        } else if (loops.length > 0) {
          map.set(i, loops[loops.length - 1].ts);
        } else {
          map.set(i, turn.ts);
        }
      } else if (t === 'assistant') {
        if (loopI < firstTokens.length) {
          map.set(i, firstTokens[loopI].ts);
        } else if (loopI < loops.length) {
          map.set(i, loops[loopI].ts);
        } else {
          map.set(i, turn.ts);
        }
        // Advance the agent loop after the assistant (1:1 with loop/first_token).
        loopI++;
      }
      // tool_result / backend_tool_call: not stored as message rows; skip.
    }
  }

  return map;
}

function clampIndex(v: number, lo: number, hi: number): number {
  if (!Number.isFinite(v)) return lo;
  return Math.max(lo, Math.min(hi, Math.floor(v)));
}

/** Parse events.jsonl into ordered event lines (invalid lines skipped). */
export function parseGrokEvents(text: string): GrokEventLine[] {
  const out: GrokEventLine[] = [];
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const o = JSON.parse(trimmed) as Record<string, unknown>;
      const type = typeof o.type === 'string' ? o.type : '';
      const ts = typeof o.ts === 'string' ? o.ts : '';
      if (!type || !ts) continue;
      const ev: GrokEventLine = { type, ts };
      if (typeof o.conversation_message_count === 'number') {
        ev.conversation_message_count = o.conversation_message_count;
      }
      if (typeof o.turn_number === 'number') {
        ev.turn_number = o.turn_number;
      }
      if (typeof o.loop_index === 'number') {
        ev.loop_index = o.loop_index;
      }
      out.push(ev);
    } catch {
      /* skip */
    }
  }
  return out;
}

/** Parse signals.json for session-level token aggregates. */
export function parseGrokSignals(text: string): GrokSignals | null {
  try {
    const o = JSON.parse(text) as Record<string, unknown>;
    const contextTokensUsed =
      typeof o.contextTokensUsed === 'number'
        ? o.contextTokensUsed
        : typeof o.context_tokens_used === 'number'
          ? o.context_tokens_used
          : 0;
    const contextWindowTokens =
      typeof o.contextWindowTokens === 'number'
        ? o.contextWindowTokens
        : typeof o.context_window_tokens === 'number'
          ? o.context_window_tokens
          : 0;
    if (contextTokensUsed <= 0 && contextWindowTokens <= 0) return null;
    return { contextTokensUsed, contextWindowTokens };
  } catch {
    return null;
  }
}

/** Collect `type` for each non-empty chat_history line (absolute index order). */
export function collectChatLineTypes(chatText: string): string[] {
  const types: string[] = [];
  for (const line of chatText.split('\n')) {
    if (!line.trim()) continue;
    try {
      const o = JSON.parse(line) as { type?: unknown };
      types.push(typeof o.type === 'string' ? o.type : 'unknown');
    } catch {
      types.push('unknown');
    }
  }
  return types;
}

/**
 * Load sibling sidecars for a chat_history path and apply timestamps + tokens
 * via the shared SessionTokenApi (TS writer).
 */
export function applyGrokSidecars(
  fileService: FileService,
  chatHistoryFile: string,
  sessionId: string,
  api: SessionTokenApi,
  opts?: { fallbackCreated?: string | null; lastAssistantIndex?: number | null },
): void {
  const sessionDir = path.dirname(chatHistoryFile);
  let lineTypes: string[];
  try {
    lineTypes = collectChatLineTypes(fileService.readFileSync(chatHistoryFile));
  } catch {
    return;
  }

  let events: GrokEventLine[] = [];
  try {
    events = parseGrokEvents(fileService.readFileSync(path.join(sessionDir, EVENTS_FILE)));
  } catch {
    /* no events */
  }

  const fallback = opts?.fallbackCreated ?? null;
  const tsMap = buildTimestampMap(lineTypes, events, fallback);
  for (const [msgIndex, ts] of tsMap) {
    const t = lineTypes[msgIndex];
    if (t === 'system' || t === 'user' || t === 'assistant' || t === 'reasoning') {
      api.updateMessageTimestamp?.(sessionId, msgIndex, ts);
    }
  }

  let signals: GrokSignals | null = null;
  try {
    signals = parseGrokSignals(fileService.readFileSync(path.join(sessionDir, SIGNALS_FILE)));
  } catch {
    /* no signals */
  }
  if (!signals || signals.contextTokensUsed <= 0) return;

  let lastAssistant = opts?.lastAssistantIndex ?? null;
  if (lastAssistant == null) {
    for (let i = lineTypes.length - 1; i >= 0; i--) {
      if (lineTypes[i] === 'assistant') {
        lastAssistant = i;
        break;
      }
    }
  }
  if (lastAssistant == null) return;

  api.updateMessageTokens(sessionId, lastAssistant, {
    inputTokens: signals.contextTokensUsed,
    outputTokens: 0,
    cacheCreationTokens: 0,
    cacheReadTokens: 0,
  });
  api.setSessionTokensEstimated(sessionId, true);
}

File: crates/spaghetti-napi/src/grok/sidecars.rs
//! Grok sidecar enrichment — timestamps (events.jsonl) + session tokens (signals.json).
//!
//! Behaviour-aligned with `packages/sdk/src/sources/grok/sidecars.ts`.

use std::collections::HashMap;
use std::path::Path;

use serde_json::Value;

/// Absolute chat_history line index → ISO timestamp.
pub type TimestampMap = HashMap<u32, String>;

#[derive(Debug, Clone, Default)]
pub struct GrokSignals {
    pub context_tokens_used: u64,
}

/// Parse events.jsonl text into (type, ts, conversation_message_count?) triples.
fn parse_events(text: &str) -> Vec<(String, String, Option<u32>)> {
    let mut out = Vec::new();
    for line in text.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let Ok(v) = serde_json::from_str::<Value>(trimmed) else {
            continue;
        };
        let ty = v.get("type").and_then(Value::as_str).unwrap_or("").to_owned();
        let ts = v.get("ts").and_then(Value::as_str).unwrap_or("").to_owned();
        if ty.is_empty() || ts.is_empty() {
            continue;
        }
        let count = v
            .get("conversation_message_count")
            .and_then(Value::as_u64)
            .map(|n| n as u32);
        out.push((ty, ts, count));
    }
    out
}

/// Collect `type` for each non-empty chat_history line.
pub fn collect_line_types(chat_text: &str) -> Vec<String> {
    let mut types = Vec::new();
    for line in chat_text.lines() {
        if line.trim().is_empty() {
            continue;
        }
        match serde_json::from_str::<Value>(line) {
            Ok(v) => types.push(
                v.get("type")
                    .and_then(Value::as_str)
                    .unwrap_or("unknown")
                    .to_owned(),
            ),
            Err(_) => types.push("unknown".into()),
        }
    }
    types
}

/// Build absolute-line-index → timestamp map (mirrors TS `buildTimestampMap`).
pub fn build_timestamp_map(
    line_types: &[String],
    events_text: &str,
    fallback_created: Option<&str>,
) -> TimestampMap {
    let events = parse_events(events_text);
    let mut map: TimestampMap = HashMap::new();

    // 1) turn_started with conversation_message_count → user at that index.
    for (ty, ts, count) in &events {
        if ty != "turn_started" {
            continue;
        }
        if let Some(mut idx) = *count {
            while (idx as usize) < line_types.len() && line_types[idx as usize] != "user" {
                idx += 1;
            }
            if (idx as usize) < line_types.len() {
                map.entry(idx).or_insert_with(|| ts.clone());
            }
        }
    }

    let leftover_turns: Vec<&str> = events
        .iter()
        .filter(|(t, _, _)| t == "turn_started")
        .map(|(_, ts, _)| ts.as_str())
        .collect();
    let stamped_users = map
        .keys()
        .filter(|&&i| line_types.get(i as usize).map(|s| s.as_str()) == Some("user"))
        .count();
    let mut turn_i = stamped_users.min(leftover_turns.len());

    let loop_starts: Vec<&str> = events
        .iter()
        .filter(|(t, _, _)| t == "loop_started")
        .map(|(_, ts, _)| ts.as_str())
        .collect();
    let first_tokens: Vec<&str> = events
        .iter()
        .filter(|(t, _, _)| t == "first_token")
        .map(|(_, ts, _)| ts.as_str())
        .collect();
    let mut loop_i = 0usize;
    let mut first_i = 0usize;

    for (i, t) in line_types.iter().enumerate() {
        let idx = i as u32;
        if map.contains_key(&idx) {
            continue;
        }
        match t.as_str() {
            "system" => {
                if let Some(fb) = fallback_created {
                    map.insert(idx, fb.to_owned());
                }
            }
            "user" => {
                if turn_i < leftover_turns.len() {
                    map.insert(idx, leftover_turns[turn_i].to_owned());
                    turn_i += 1;
                } else if let Some(fb) = fallback_created {
                    map.insert(idx, fb.to_owned());
                }
            }
            "reasoning" => {
                if loop_i < loop_starts.len() {
                    map.insert(idx, loop_starts[loop_i].to_owned());
                    loop_i += 1;
                } else if first_i < first_tokens.len() {
                    map.insert(idx, first_tokens[first_i].to_owned());
                    first_i += 1;
                }
            }
            "assistant" => {
                if first_i < first_tokens.len() {
                    map.insert(idx, first_tokens[first_i].to_owned());
                    first_i += 1;
                }
            }
            _ => {}
        }
    }

    map
}

pub fn parse_signals(text: &str) -> Option<GrokSignals> {
    let v: Value = serde_json::from_str(text).ok()?;
    let used = v
        .get("contextTokensUsed")
        .or_else(|| v.get("context_tokens_used"))
        .and_then(Value::as_u64)
        .unwrap_or(0);
    if used == 0 {
        return None;
    }
    Some(GrokSignals {
        context_tokens_used: used,
    })
}

/// Load timestamp map + signals from sibling files next to chat_history.
pub fn load_sidecars(
    chat_history: &Path,
    fallback_created: Option<&str>,
) -> (TimestampMap, Option<GrokSignals>, Vec<String>) {
    let session_dir = match chat_history.parent() {
        Some(p) => p,
        None => return (HashMap::new(), None, Vec::new()),
    };

    let chat_text = std::fs::read_to_string(chat_history).unwrap_or_default();
    let line_types = collect_line_types(&chat_text);

    let events_text =
        std::fs::read_to_string(session_dir.join("events.jsonl")).unwrap_or_default();
    let ts_map = build_timestamp_map(&line_types, &events_text, fallback_created);

    let signals = std::fs::read_to_string(session_dir.join("signals.json"))
        .ok()
        .and_then(|t| parse_signals(&t));

    (ts_map, signals, line_types)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn assigns_turn_and_token_timestamps() {
        let types = vec![
            "system".into(),
            "user".into(),
            "reasoning".into(),
            "assistant".into(),
            "tool_result".into(),
            "assistant".into(),
        ];
        let events = r#"
{"ts":"2026-04-01T10:00:00.000Z","type":"turn_started","conversation_message_count":1}
{"ts":"2026-04-01T10:00:01.000Z","type":"loop_started"}
{"ts":"2026-04-01T10:00:02.000Z","type":"first_token"}
{"ts":"2026-04-01T10:00:05.000Z","type":"first_token"}
"#;
        let map = build_timestamp_map(&types, events, Some("2026-04-01T09:00:00.000Z"));
        assert_eq!(map.get(&0).map(String::as_str), Some("2026-04-01T09:00:00.000Z")); // system fallback
        assert_eq!(map.get(&1).map(String::as_str), Some("2026-04-01T10:00:00.000Z")); // user
        assert_eq!(map.get(&2).map(String::as_str), Some("2026-04-01T10:00:01.000Z")); // reasoning
        assert_eq!(map.get(&3).map(String::as_str), Some("2026-04-01T10:00:02.000Z")); // assistant
        assert_eq!(map.get(&5).map(String::as_str), Some("2026-04-01T10:00:05.000Z")); // 2nd assistant
        assert!(!map.contains_key(&4)); // tool_result skipped
    }

    #[test]
    fn parses_signals_context_tokens() {
        let s = parse_signals(r#"{"contextTokensUsed":106352,"contextWindowTokens":500000}"#).unwrap();
        assert_eq!(s.context_tokens_used, 106352);
    }
}
